File: exercises-java/src/main/java/nl/trivento/propertybased/exercise2/TransitiveDependencies.java
package nl.trivento.propertybased.exercise2;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class TransitiveDependencies {


    /**
     *
     */
    public static Set<Character> findTransitiveDependencies(char key, Map<Character, Set<Character>> dependencies) {
        return findTransitiveDependencies(key, new HashSet<>(), dependencies);
    }

    private static Set<Character> findTransitiveDependencies(char currentKey, Set<Character> acc, Map<Character, Set<Character>> dependencies) {
        if (dependencies.containsKey(currentKey)) {
            Set<Character> directDependencies = dependencies.get(currentKey);

            Set<Character> indirectDependencies = new HashSet<>();
            for (char key: directDependencies) {
                indirectDependencies.addAll(findTransitiveDependencies(key, directDependencies, dependencies));
            }
            indirectDependencies.addAll(directDependencies);
            return indirectDependencies;
        } else {
            return acc;
        }
    }
}

File: exercises-scala/src/main/scala/nl/trivento/propertybased/exercise2/TransitiveDependencies.scala
package nl.trivento.propertybased.exercise2

/**
  * The findTransitiveDependencies function should calculate the full set of dependencies.
  * One of the insidious things about dependencies is that they are transitive.
  * Meaning that, if 'A' depends on 'B' and 'B' depends on 'C', then 'A' also depends on 'C'.
  *
  * Below you see an example:
  *
  * Map('A' -> Set('B', 'C'),
  *     'B' -> Set('C', 'E'),
  *     'C' -> Set('G'),
  *     'D' -> Set('A', 'F'),
  *     'E' -> Set('F'),
  *     'F' -> Set('H'))
  *
  * 'A' has two direct dependencies, namely 'B' and 'C'.
  * Next to that it has the following indirect dependencies: 'E', 'F', 'G' and 'H'
  *
  * So the full set of dependencies for 'A' are 'B', 'C', 'E', 'F', 'G' and 'H'
  *
  */
object TransitiveDependencies {


  def findTransitiveDependencies(key: Char, dependencies: Map[Char, Set[Char]]): Set[Char] = {
    def loop(key: Char, acc: Set[Char]): Set[Char] = {
      dependencies.get(key) match {
        case Some(directDependencies) if directDependencies.diff(acc).isEmpty => acc
        case Some(directDependencies) =>
          val indirectDependencies: Set[Char] = directDependencies.flatMap(key => loop(key, acc ++ directDependencies))
          directDependencies ++ indirectDependencies
        case None => acc
      }
    }
    loop(key, acc = Set.empty)
  }
}
